doc/conf.py

```diff
@@ -767,7 +767,52 @@ def generate_min_dependency_substitutions(app):
 
     output = output.getvalue()
 
-    with (Path(".") / "min_dependency_substitutions.rst").open("w") as f:
+    with (Path(".") / "min_dependency_substitutions.rst").open(
+        "w", encoding="utf-8"
+    ) as f:
+        f.write(output)
+
+
+def generate_index_rst():
+    """Generate index.rst to imply the overall structure of docs."""
```

Comment:
```suggestion
    """Generate index.rst to simplify the overall structure of docs."""
```
Feels like a word/letters went missing here. This is my guess of what it was meant to be

Reply:
What I wanted to say is: `index.rst` is the `root_doc` so its toctree will determine what will show up in the topbar. That's what I meant by "imply the overall structure". Do you have an alternative expression in mind?

Reply:
I'm not sure :-/ Can you explain again why the `index.rst` is generated instead of being a plain old file? Is it because we need to be able to generate/replace `development_link`?

I'd put the reason for why we need to generate the page in the docstring. Maybe you already did that?

Reply:
I think I put it somewhere in a comment of this PR but forgot to put it in the file 😅

Reply:
`rst_pro/epilog` and some googling got me to https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#substitutions

That seems like it would do the trick here? But I'm not really a rST/sphinx expert when it comes to the pros and cons of the two options

Reply:
It seems that `rst` does not support nested inline markup. In particular, `Development <|development-link|>` won't work - it will link to literally `|development-link|` which does not exist. Let `development-ref` be `development-link` with the angular brackets and do `Development |development-ref|` will not work either because it will treat literally `Development |development-ref|` as a document reference which clearly does not exist again.

I'm also not an expert in these... all based on googling and trials and errors.

Reply:
Let's see if someone has a smart idea/wisdom. Otherwise we'll go with the template approach you proposed.

Reply:
Well there is indeed one alternative: we have the configuration `"navbar_center": ["navbar-nav"]` which is a component provided by `pydata-sphinx-theme`. We can write our own HTML component and put it here instead, and with HTML I believe we can have access to the context variables (added in `html_context`). One side benefit is that, we do not need the doc structure (defined by toctree) and the top bar (with this approach, manually created with HTML) to be in the same order, which is originally not the case.

Reply:
Having looked at d9b71322fbceed260fb0aefc5467df2c914b5d73 I think I agree with you about it being more complicated. At least I think if we came back in three months and saw the template approach we'd quickly understand how it works, why it is there and how to fix/modify it. The approach with `generate_header_nav_html` might be more "proper" but I think for someone who knows nothing about sphinx and the website it is harder to understand why it is there, how it works and how to fix/modify it.

Reply:
Thanks for your reply, so I will revert to the previous approach for now.

The diff of the alternative approach in case someone wants to take a look: [`new_web_theme..1563518`](https://github.com/scikit-learn/scikit-learn/compare/new_web_theme..1563518)